spring-rabbit/src/main/java/org/springframework/amqp/rabbit/connection/CachingConnectionFactory.java

```diff
@@ -1005,45 +1074,50 @@ public Channel createChannel(boolean transactional) {
 		public void close() {
 			if (CachingConnectionFactory.this.cacheMode == CacheMode.CONNECTION) {
 				synchronized (CachingConnectionFactory.this.connectionMonitor) {
-					if (!this.target.isOpen()
-							|| CachingConnectionFactory.this.idleConnections.size() >=
-									CachingConnectionFactory.this.connectionCacheSize) {
-						if (logger.isDebugEnabled()) {
-							logger.debug("Completely closing connection '" + this + "'");
-						}
-						if (this.target.isOpen()) {
-							RabbitUtils.closeConnection(this.target);
-						}
-						this.notifyCloseIfNecessary();
-						CachingConnectionFactory.this.openConnections.remove(this);
-						CachingConnectionFactory.this.openConnectionNonTransactionalChannels.remove(this);
-						CachingConnectionFactory.this.openConnectionTransactionalChannels.remove(this);
-					}
-					else {
-						if (!CachingConnectionFactory.this.idleConnections.contains(this)) {
+					/*
+					 * Only connectionCacheSize open idle connections are allowed.
+					 */
+					if (!CachingConnectionFactory.this.idleConnections.contains(this)) {
+						if (!this.target.isOpen()
+								|| countOpenIdleConnections() >= CachingConnectionFactory.this.connectionCacheSize) {
 							if (logger.isDebugEnabled()) {
-								logger.debug("Returning connection '" + this + "' to cache");
-							}
-							CachingConnectionFactory.this.idleConnections.add(this);
-							if (CachingConnectionFactory.this.connectionHighWaterMark
-									.get() < CachingConnectionFactory.this.idleConnections.size()) {
-								CachingConnectionFactory.this.connectionHighWaterMark
-										.set(CachingConnectionFactory.this.idleConnections.size());
+								logger.debug("Completely closing connection '" + this + "'");
 							}
+							destroy();
```

Comment:
Gary, is it fine that we go ahead in this method after the `destroy()` and say in logs like:
```
2016-03-22 15:42:29,792 DEBUG Test worker [org.springframework.amqp.rabbit.connection.CachingConnectionFactory] - <Completely closing connection 'Proxy@46204d1 Dedicated Rabbit Connection: SimpleConnection@261b168a [delegate=amqp://guest@127.0.0.1:5672/, localPort= 55310]'>
2016-03-22 15:42:29,794 DEBUG Test worker [org.springframework.amqp.rabbit.connection.CachingConnectionFactory] - <Returning connection 'Proxy@46204d1 Dedicated Rabbit Connection: null' to cache>
```
?
Shouldn't we `return` from here altogether?